### PR TITLE
Core: read delete files in parallel

### LINKED ISSUE #3118: Read delete files in parallel.
In upsert/cdc case, we usually will get a lot of pos-delete and eq-delete files. When we read/rewrite data from the v2 table, `DeleteFilter` will open all referenced pos-delete files and eq-delete files for each data file to construct the posDeleteSet and eqDeleteSet. 

Currently, that all work will handled by same thread for each `CombinedScanTask` and all delete files are read serially, that is mean iceberg read a delete file must wait for the last file to be read and `DeleteFilter` will take a lot of time to open and read delete files. 
I think `DeleteFilter` should read delete files in parallel when construct the posDeleteSet and eqDeleteSet to speed up reading v2 table.

### PR DESCRIPTION
**Pull request by Reo-LEI**
This PR is trying to close #3118.

I move `ParallelIterable` to api module and use it impl parallel `concat` api in `CloseableIterable`. Now we could use the parallel `concat` to replace the serial `concat` and read delete files in parallel when `DeleteFilter` construct the equalitySet and positionSet.

### CODE REVIEW

**Comment by Reo-LEI** [comment]
@rdblue @aokolnychyi @kbendick @openinx @stevenzwu Could you take a look of this? :)

File: core/src/main/java/org/apache/iceberg/util/ThreadPools.java [hunk 1]
```diff
@@ -61,6 +61,15 @@ public static ExecutorService getWorkerPool() {
     return WORKER_POOL;
   }
 
+  /**
+   * Return parallelism for "worker" thread-pool.
+   * In default, we submit 2 tasks per worker at a time.
+   * @return the parallelism of the worker pool.
+   */
+  public static int getPoolParallelism() {
```
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/util/ThreadPools.java:69]: I am wondering if it is better to define a constant as 
```
public static final int DEFAULT_WORKER_THREAD_POOL_PARALLELISM = WORKER_THREAD_POOL_SIZE * 2;
```

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 2]
```diff
@@ -135,8 +136,10 @@ protected long pos(T record) {
       Iterable<CloseableIterable<Record>> deleteRecords = Iterables.transform(deletes,
           delete -> openDeletes(delete, deleteSchema));
       StructLikeSet deleteSet = Deletes.toEqualitySet(
-          // copy the delete records because they will be held in a set
-          CloseableIterable.transform(CloseableIterable.concat(deleteRecords), Record::copy),
+          CloseableIterable.transform(
+              CloseableIterable.concat(deleteRecords, ThreadPools.getWorkerPool(), ThreadPools.getPoolParallelism()),
```
Comment (stevenzwu) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:140]: did this PR actually change or increase the parallelism? it seems to me we are still using the default parallelism of `2 * ThreadPools.WORKER_THREAD_POOL_SIZE`.  Is the purpose of this PR mainly refactoring the API and supporting different parallelism (if needed in the future)?

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 3]
```diff
@@ -205,14 +212,12 @@ protected boolean shouldKeep(T item) {
       case AVRO:
         return Avro.read(input)
             .project(deleteSchema)
-            .reuseContainers()
```
Comment (stevenzwu) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:208]: is this related?

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java [hunk 4]
```diff
@@ -117,6 +118,12 @@ public O next() {
     };
   }
 
+  static <E> CloseableIterable<E> concat(Iterable<? extends Iterable<E>> iterable,
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:121]: This doesn't concatenate the iterables, it mixes them together, so we should come up with a better name. Maybe `combine`?

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 5]
```diff
@@ -72,8 +73,7 @@ private ParallelIterator(Iterable<? extends Iterable<T>> iterables,
             }
           }).iterator();
       this.workerPool = workerPool;
-      // submit 2 tasks per worker at a time
-      this.taskFutures = new Future[2 * ThreadPools.WORKER_THREAD_POOL_SIZE];
+      this.taskFutures = new Future[parallelism];
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:76]: Is it possible to base this on the size of the `ExecutorService`, rather than pass it in?

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 3 above)
Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:208]: I think this is needed because records are now placed on a queue inside the parallel iterator. Reusing the record instance instead of copying would cause a problem.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 6]
```diff
@@ -189,7 +192,11 @@ protected boolean shouldKeep(T item) {
     if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
       return Deletes.filter(
           records, this::pos,
-          Deletes.toPositionSet(dataFile.path(), CloseableIterable.concat(deletes)));
+          Deletes.toPositionSet(
+              dataFile.path(),
+              CloseableIterable.concat(deletes, ThreadPools.getWorkerPool(), ThreadPools.getPoolParallelism())
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:197]: I think that parallelizing this requires adding an option. There are many cases where we wouldn't want to parallelize and cases where we want to use a thread pool besides the shared worker pool. It would be better to pass a pool in and parallelize if it is non-null.

Comment (jackye1995) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:197]: +1, something like `read.delete.position.num-threads` table property can work.

In general we follow this pattern when configuring and creating thread pools and executor service: https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java#L187-L194

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 2 above)
Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:140]: The purpose of this PR is to be able to read deleteFile in parallel, before this PR, iceberg will read all delete files in one thread  serially. Now, delete files will be read in parallel by a worker pool.

File: core/src/main/java/org/apache/iceberg/util/ThreadPools.java (hunk 1 above)
Comment (Reo-LEI) [reply core/src/main/java/org/apache/iceberg/util/ThreadPools.java:69]: LGTM, I will change that as a constant.

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java (hunk 4 above)
Comment (Reo-LEI) [reply api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:121]: Agree! I prefer `combine`, because user can set `executorService` to null to run task serially.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java (hunk 5 above)
Comment (Reo-LEI) [reply api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:76]: I wanna to do this at first, but `ExecutorService` does not seem to provide the API to get the pool size. I think the reason is only part of `ExecutorService` will concern the pool size, and such as pool size of `CachedThreadPool` will be `Integer.MAX_VALUE`.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 6 above)
Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:197]: LGTM, I will trying to make it configurable.

Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java]: I add an optional method to set the read pool and config the read parallelism by code.  @rdblue 
And I think we could config the read parallelism by 'iceberg.worker.num-threads' rather than table property, because I don't want to spread the table property widely.  @jackye1995

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 7]
```diff
@@ -90,6 +95,9 @@ protected DeleteFilter(FileScanTask task, Schema tableSchema, Schema requestedSc
     this.eqDeletes = eqDeleteBuilder.build();
     this.requiredSchema = fileProjection(tableSchema, requestedSchema, posDeletes, eqDeletes);
     this.posAccessor = requiredSchema.accessorForField(MetadataColumns.ROW_POSITION.fieldId());
+
+    this.readService = ThreadPools.getWorkerPool();
+    this.readParallelism = ThreadPools.WORKER_THREAD_POOL_PARALLELISM;
```
Comment (kbendick) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:100]: Should the default implementation use the worker thread pool? It seems if we make it configurable, that we would want to check the configuration first as a parallelism of 1 might not require the worker pool at all (like the current behavior).

Though possibly I missed something when looking through it that makes the readService always required, even when the user provided delete file read parallelism = 1.

And I'm still in favor of instantiating a named thread pool like Jack mentioned, like here: https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java#L187-L194. Unless there's a reason not to do that.

Using a named thread pool makes debugging much easier.

Comment (kbendick) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:100]: Thinking about it further, it's probably keeps the code easier to read to keep an executor service in place even in the case that somebody uses 1.

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java [hunk 8]
```diff
@@ -117,6 +118,14 @@ public O next() {
     };
   }
 
+  static <E> CloseableIterable<E> combine(Iterable<? extends Iterable<E>> iterable,
+                                          ExecutorService executorService,
+                                          int parallelism) {
+    Preconditions.checkNotNull(executorService, "ExecutorService shouldn't be null");
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:124]: The error message should be "Invalid executor service: null".

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java [hunk 9]
```diff
@@ -117,6 +118,14 @@ public O next() {
...
+    Preconditions.checkArgument(parallelism > 0, "Parallelism must be > 0");
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:125]: Here as well, I think the error message should follow our conventions to state the problem, give context, and then show the bad value: "Invalid parallelism (not positive): %s"

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java [hunk 10]
```diff
@@ -117,6 +118,14 @@ public O next() {
...
+    return new ParallelIterable<>(iterable, executorService, parallelism);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:126]: If parallelism is 1, this should just call `concat` right?

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 11]
```diff
@@ -28,25 +28,25 @@
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
 import org.apache.iceberg.exceptions.RuntimeIOException;
-import org.apache.iceberg.io.CloseableGroup;
-import org.apache.iceberg.io.CloseableIterable;
-import org.apache.iceberg.io.CloseableIterator;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
 
 public class ParallelIterable<T> extends CloseableGroup implements CloseableIterable<T> {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:34]: Moving this class is a breaking change. I'd be okay with it if the purpose was to make it private (and there was an appropriate deprecation in its place). But since this isn't private, I'm not sure the purpose or value of moving it.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java (hunk 5 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:76]: What about `ThreadPoolExecutor#getPoolSize()`?

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:76]: This is okay, then. But it still needs to double the number of threads in the pool. Can you add back the multipler?

File: core/src/main/java/org/apache/iceberg/util/ThreadPools.java [hunk 12]
```diff
@@ -38,6 +38,9 @@ private ThreadPools() {
       WORKER_THREAD_POOL_SIZE_PROP,
       Runtime.getRuntime().availableProcessors());
 
+  // In default, we submit 2 tasks per worker at a time.
+  public static final int WORKER_THREAD_POOL_PARALLELISM = WORKER_THREAD_POOL_SIZE * 2;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/ThreadPools.java:42]: This doesn't make sense here and complicates the `ParallelIterator` API. I think the parallel iterator should just double the parallelism and document that it expects the parallelism to be the pool size.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 13]
```diff
@@ -90,6 +95,9 @@ protected DeleteFilter(FileScanTask task, Schema tableSchema, Schema requestedSc
     this.eqDeletes = eqDeleteBuilder.build();
     this.requiredSchema = fileProjection(tableSchema, requestedSchema, posDeletes, eqDeletes);
     this.posAccessor = requiredSchema.accessorForField(MetadataColumns.ROW_POSITION.fieldId());
+
+    this.readService = ThreadPools.getWorkerPool();
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:99]: This should not use the worker pool. That is for job planning.

Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:99]: If this is just for a default, then the default should be the existing behavior with no executor service and `CloseableIterable.concat`.

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java (hunk 8 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:124]: You may want to allow null if the parallelism is 1. That will make the calling code much easier.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 3 above)
Comment (rdblue) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:208]: This should only be done if the parallelism is greater than 1.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 14]
```diff
@@ -135,8 +149,10 @@ protected long pos(T record) {
       Iterable<CloseableIterable<Record>> deleteRecords = Iterables.transform(deletes,
           delete -> openDeletes(delete, deleteSchema));
       StructLikeSet deleteSet = Deletes.toEqualitySet(
-          // copy the delete records because they will be held in a set
-          CloseableIterable.transform(CloseableIterable.concat(deleteRecords), Record::copy),
+          CloseableIterable.transform(
+              CloseableIterable.combine(deleteRecords, readService, readParallelism),
+              deleteRecord -> (StructLike) deleteRecord
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:154]: Is the outer `transform` call needed? The iterable should be of `Record` which is already `StructLike`.

**Comment by Reo-LEI** [comment]
I have been make the parallelize in configurable and let it run in a separate thread pool. 
Could you help review this PR again? @rdblue @stevenzwu @kbendick @jackye1995

**Comment by rdblue** [comment]
Thanks, @Reo-LEI. I'll take another look.

File: api/src/main/java/org/apache/iceberg/io/CloseableIterable.java [hunk 15]
```diff
@@ -117,6 +118,16 @@ public O next() {
     };
   }
 
+  static <E> CloseableIterable<E> combine(Iterable<CloseableIterable<E>> iterable,
+                                          ExecutorService workerPool,
+                                          int workerPoolSize) {
+    if (workerPool == null) {
+      return concat(iterable);
+    }
+    Preconditions.checkArgument(workerPoolSize > 0, "Invalid workerPoolSize (not positive): " + workerPoolSize);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/CloseableIterable.java:127]: Style: in this project, we use empty lines to separate control flow blocks from the following statements.

I'd also add a new line after the precondition.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 16]
```diff
@@ -72,8 +73,8 @@ private ParallelIterator(Iterable<? extends Iterable<T>> iterables,
             }
           }).iterator();
       this.workerPool = workerPool;
-      // submit 2 tasks per worker at a time
-      this.taskFutures = new Future[2 * ThreadPools.WORKER_THREAD_POOL_SIZE];
+      // In default, we submit 2 tasks per worker at a time.
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:76]: Style: please remove personal pronouns. Using "we" or "I" makes documentation longer and less direct. Also, no need for punctuation in line comments.

Actually, I don't think that this comment needs to change at all. It looks perfectly fine without the changes.

File: core/src/main/java/org/apache/iceberg/AllDataFilesTable.java [hunk 17]
```diff
@@ -128,9 +127,9 @@ public long targetSplitSize() {
   }
 
   private static CloseableIterable<ManifestFile> allDataManifestFiles(List<Snapshot> snapshots) {
-    try (CloseableIterable<ManifestFile> iterable = new ParallelIterable<>(
-        Iterables.transform(snapshots, snapshot -> (Iterable<ManifestFile>) () -> snapshot.dataManifests().iterator()),
-        ThreadPools.getWorkerPool())) {
+    try (CloseableIterable<ManifestFile> iterable = CloseableIterable.combine(
+        Iterables.transform(snapshots, snapshot -> CloseableIterable.withNoopClose(snapshot.dataManifests())),
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/AllDataFilesTable.java:131]: I don't think that this should add a noop close here. Can you avoid adding this in all the updated calls to create a parallel iterable?

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java [hunk 18]
```diff
@@ -77,8 +77,9 @@ private Deletes() {
     return filter.filter(rows);
   }
 
-  public static StructLikeSet toEqualitySet(CloseableIterable<StructLike> eqDeletes, Types.StructType eqType) {
-    try (CloseableIterable<StructLike> deletes = eqDeletes) {
+  public static <T extends StructLike> StructLikeSet toEqualitySet(CloseableIterable<T> eqDeletes,
+                                                                   Types.StructType eqType) {
+    try (CloseableIterable<T> deletes = eqDeletes) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/deletes/Deletes.java:82]: Why was this change needed?

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 19]
```diff
@@ -67,7 +74,11 @@
   private final Schema requiredSchema;
   private final Accessor<StructLike> posAccessor;
 
-  protected DeleteFilter(FileScanTask task, Schema tableSchema, Schema requestedSchema) {
+  private int readDeletesWorkerPoolSize;
+  private ExecutorService readDeletesService;
+
+  protected DeleteFilter(FileScanTask task, Schema tableSchema, Schema requestedSchema,
+                         Map<String, String> tableProperties) {
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:81]: This class should not take table properties. I think it should take an executor service instead.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 20]
```diff
@@ -108,6 +123,15 @@ protected long pos(T record) {
...
   }
 
+  private ExecutorService readDeletesService(int workerPoolSize) {
+    return MoreExecutors.getExitingExecutorService(
+        (ThreadPoolExecutor) Executors.newFixedThreadPool(
+            workerPoolSize,
+            new ThreadFactoryBuilder()
+                .setNameFormat("Read-delete-Service-%d")
+                .build()));
+  }
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:133]: When is this executor service cleaned up?

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 21]
```diff
@@ -123,6 +123,9 @@ private TableProperties() {
   public static final String SPLIT_OPEN_FILE_COST = "read.split.open-file-cost";
   public static final long SPLIT_OPEN_FILE_COST_DEFAULT = 4 * 1024 * 1024; // 4MB
 
+  public static final String READ_DELETE_FILES_WORKER_POOL_SIZE = "read.deletes.num-threads";
+  public static final int READ_DELETE_FILES_WORKER_POOL_SIZE_DEFAULT = 1;  // read delete files in serial.
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:127]: I don't think this makes much sense as a table property. Table properties are for table configuration, but this is an engine concern. Many engines handle parallelism internally so this wouldn't be appropriate. I think that Flink should expose a setting and manage the thread pool for using a delete file reader pool.

Comment (Reo-LEI) [reply core/src/main/java/org/apache/iceberg/TableProperties.java:127]: I admit that this configuration may not be appropriate as a table property. But I think this optimization should apply to all engines, not only flink. Such as I will sync mysql cdc data to iceberg by flink and rewrite and merge the delete files into data files by spark. Maybe we chould config this through `SystemProperties` like we config `iceberg.worker.num-threads`.
https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/SystemProperties.java#L35

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 20 above)
Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:133]: I think we chould create a custom executor and set `corePoolSize = 0`, `maximumPoolSize = poolSzie`, `workQueue = new LinkedBlockingQueue<Runnable>()` and let executor shutdown automatically. Otherwise, we need to call `shutdown` in `filter` or add `close` metod and call by `DeleterFilter` caller.

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java (hunk 18 above)
Comment (Reo-LEI) [reply core/src/main/java/org/apache/iceberg/deletes/Deletes.java:82]: To avoid to transform `Record` to `StructLike` as your comment https://github.com/apache/iceberg/pull/3120#discussion_r711824192.

File: core/src/main/java/org/apache/iceberg/AllDataFilesTable.java (hunk 17 above)
Comment (Reo-LEI) [reply core/src/main/java/org/apache/iceberg/AllDataFilesTable.java:131]: I think I can't avoid adding this, becasue `combine` will call `concat` when `workerPool` is null, and `concat` receive `Iterable<CloseableIterable<E>>` but not `Iterable<Iterable<E>>`. So I need to add a noop close to wrap it and pass to `combine`.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 20 above)
Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java]: I make this executor service as static. Now, different filter task can share the same executor service, and we don't need to consider when we should shutdown and cleaned up the executor service.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java (hunk 19 above)
Comment (Reo-LEI) [reply data/src/main/java/org/apache/iceberg/data/DeleteFilter.java]: I think `DeleteFilter` should make sure the reader will read delete files in parallel by itself, if we config it, but not pass a executor service by the caller. So I create the read service internally, and read the config from system properties.

File: core/src/main/java/org/apache/iceberg/TableProperties.java (hunk 21 above)
Comment (Reo-LEI) [reply core/src/main/java/org/apache/iceberg/TableProperties.java]: I move this config to `SystemProperties`. So we can stop propagate the 'tableProperties' and handle the parallelism internally.

**Comment by Reo-LEI** [comment]
I adressed some comment and leave some comment. @rdblue Could you take another looks of this PR? 😄

File: core/src/main/java/org/apache/iceberg/TableProperties.java (hunk 21 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/TableProperties.java]: Is it not possible to have Flink pass in the executor service it chooses to use?

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 22]
```diff
@@ -28,25 +28,39 @@
...
+ * @deprecated please use {@link CloseableIterable#combine(Iterable, ExecutorService, int)} instead.
+ */
+@Deprecated
 public class ParallelIterable<T> extends CloseableGroup implements CloseableIterable<T> {
   private final Iterable<? extends Iterable<T>> iterables;
   private final ExecutorService workerPool;
+  private final int workerPoolSize;
 
+  /**
+   * @deprecated please use {@link CloseableIterable#combine(Iterable, ExecutorService, int)} instead.
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:45]: You can omit "please" from documentation so that docs are direct and as short as possible.

Also, can you add "will be removed in 0.14.0"? We like to keep track of when things can be removed.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 23]
```diff
@@ -72,8 +87,8 @@ private ParallelIterator(Iterable<? extends Iterable<T>> iterables,
             }
           }).iterator();
       this.workerPool = workerPool;
-      // submit 2 tasks per worker at a time
-      this.taskFutures = new Future[2 * ThreadPools.WORKER_THREAD_POOL_SIZE];
+      // submit 2 tasks per worker at a time.
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:90]: Can you remove the non-functional change on this line? We don't want unnecessary changes to cause commit conflicts.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java [hunk 24]
```diff
@@ -28,25 +28,39 @@
...
-import org.apache.iceberg.io.CloseableIterable;
-import org.apache.iceberg.io.CloseableIterator;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
 
+/**
+ * Run iterables in parallel.
+ * @deprecated please use {@link CloseableIterable#combine(Iterable, ExecutorService, int)} instead.
+ */
+@Deprecated
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:38]: This class is not deprecated. It is still public and will not be removed. This should be created using `CloseableIterable#combine` rather than directly using the constructor. Can you remove this deprecation?

File: core/src/main/java/org/apache/iceberg/AllDataFilesTable.java (hunk 17 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/AllDataFilesTable.java:131]: This does need to be fixed. `ParallelIterable` was constructed using `Iterable<? extends Iterable<T>>`. There should be a version of `CloseableIterable.combine` that accepts the same type. That may mean updating `CloseableIterable.concat` to accept the same `? extends Iterable<T>` in addition to strictly a `CloseableIterable<T>`. But that should be okay, since you can update to close the iterable if it is closeable.

File: core/src/main/java/org/apache/iceberg/SystemProperties.java [hunk 25]
```diff
@@ -39,11 +39,24 @@ private SystemProperties() {
    */
   public static final String SCAN_THREAD_POOL_ENABLED = "iceberg.scan.plan-in-worker-pool";
 
-  static boolean getBoolean(String systemProperty, boolean defaultValue) {
+  /**
+   * Whether to use the shared worker pool when planning table scans.
+   */
+  public static final String READ_DELETE_FILES_WORKER_POOL_SIZE = "iceberg.worker.read-deletes-num-threads";
+
+  public static boolean getBoolean(String systemProperty, boolean defaultValue) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SystemProperties.java:47]: As I said elsewhere, I don't think that this feature should be controlled through a system property. This should be a Flin-specific property for now and we can introduce a similar config for Spark later. But since this violates Spark's threading model on executors, we don't want to make this global.

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java (hunk 18 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/deletes/Deletes.java:82]: Seems reasonable.

File: data/src/main/java/org/apache/iceberg/data/DeleteFilter.java [hunk 26]
```diff
@@ -62,6 +68,13 @@
       MetadataColumns.DELETE_FILE_PATH,
       MetadataColumns.DELETE_FILE_POS);
 
+  private static final int READ_DELETES_WORKER_POOL_SIZE_DEFAULT = 0; // read delete files in serial.
+  private static final int READ_DELETES_WORKER_POOL_SIZE = SystemProperties.getInteger(
+      SystemProperties.READ_DELETE_FILES_WORKER_POOL_SIZE, READ_DELETES_WORKER_POOL_SIZE_DEFAULT);
+  private static final ExecutorService READ_DELETES_SERVICE = READ_DELETES_WORKER_POOL_SIZE <= 1 ? null :
+      MoreExecutors.getExitingExecutorService((ThreadPoolExecutor) Executors.newFixedThreadPool(
+          READ_DELETES_WORKER_POOL_SIZE, new ThreadFactoryBuilder().setNameFormat("Read-delete-Service-%d").build()));
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:76]: It looks like the changes other than the ones in this file are to make it easier to use `ParallelIterable`. Can we separate those changes from these for Flink and `DeleteFilter`? A separate PR for the `CloseableIterable` changes would make it easier to review this when there are more Flink changes to handle the executor service and pass it into `DeleteFilter`.

File: api/src/main/java/org/apache/iceberg/io/ParallelIterable.java (hunk 24 above)
Comment (jackye1995) [reply api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:38]: I see the module here also changed. I don't think we are allowed to move public class to another module?

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/ParallelIterable.java:38]: Good catch, @jackye1995!

**Comment by Reo-LEI** [comment]
@rdblue @stevenzwu @kbendick @jackye1995 Thanks every one for review! As Ryan's [comment](https://github.com/apache/iceberg/pull/3120#discussion_r742123603), current implementation will violates Spark's threading model. And since I start work on #3323, I think we should improve the read performance by rewrite data files and delete files in time, maybe this PR is not necessary. I will do some test after I finish #3323 and see how much benefit this optimization can bring, and then consider whether we still need this optimization. Thanks every one's work again!
